### DISCUSSION TITLE
Use pycargoebuild to automate Gentoo ebuilds

### DISCUSSION THREAD

**Discussion by Interpause** [filips123/PWAsForFirefox#392] [Ideas]
Hello, it is me again. I stumbled across https://github.com/projg2/pycargoebuild, which based on https://github.com/projg2/pycargoebuild/commit/9d1a4632eb36d928cac61389cd55e2c41cd97ca5 has support for specifying Git-installed dependencies into GIT_CRATES, solving the issue that https://github.com/gentoo/cargo-ebuild had. Seems to also be maintained by one of the senior Gentoo developers. I am going to try it now, if it works, ebuild automation should be possible.

**Comment by filips123** [comment]
That would be nice. I think that `pycargoebuild` needs to be run on Gentoo to obtain some Portage files, but GitHub Actions support Docker containers, so it should be fine to run it there.

The other thing that I'm interested about is how to submit update to GURU repository. Is it possible to just automatically push the new version through Git (or something similar) from GitHub Actions? And are automatic updates even allowed for GURU?

**Reply by Interpause** [reply]
Should be allowed I think. There are two CLI tools involved in committing to GURU. `pkgcheck` checks that ebuilds adhere to standards, and `pkgdev` helps with writing compliant commits, signing them and pushing to GURU. Commits are pushed to the dev branch, after which a high-up maintainer cherrypicks them into main. Alternatively, it should be possible to get permission to push into the official gentoo repository: https://wiki.gentoo.org/wiki/Submitting_ebuilds. I don't have experience writing github actions but I can provide the commands needed when I have time.

**Comment by Interpause** [comment]
Here is a rough outline:

```sh
cd PWAsForFirefox/native

# Replace versions as per build guide
sed -i "s/version = \"0.0.0\"/version = \"$VERSION\"/g" Cargo.toml
sed -i "s/DISTRIBUTION_VERSION = '0.0.0'/DISTRIBUTION_VERSION = '$VERSION'/g" userchrome/profile/chro
me/pwa/chrome.jsm

# Generate ebuild
pycargoebuild .

# Add in other required ebuild functions
# Maybe like:
# template.sh >> firefoxpwa-x.x.x.ebuild
# sed ...
# Can refer to https://github.com/gentoo-mirror/guru/blob/master/www-plugins/firefoxpwa/firefoxpwa-2.7.3.ebuild

# For GURU, I am not sure for official Gentoo repository
mv firefoxpwa-x.x.x.ebuild guru/www-plugins/firefoxpwa
cd guru/www-plugins/firefoxpwa
ebuild firefoxpwa-x.x.x.ebuild manifest
git add -A
pkgdev commit --signoff --scan=true
pkgdev push
```

**Reply by filips123** [reply]
I'll check how I can use this in GitHub Actions. And for getting permission to push to GURU, I need to create an account on Gentoo's Bugzilla and open a bug asking for permission, right?

**Reply by filips123** [reply]
Also, does `KEYWORDS="~amd64"` mean that the package can only work on 64-bits? As I'm not that familiar with ebuilds, is there a reason for that? Can it be updated to include keywords for other architectures that are normally supported (so, at least x86, arm and arm64)?

**Reply by filips123** [reply]
They don't want to give push permissions to third-party services such as GitHub Actions runners for security reasons (which makes sense). Because of this, I will probably set up automatic PRs to GURU repository on GitHub (but also try to get normal push permission if I need to fix something manually).

**Reply by Interpause** [reply]
> Also, does `KEYWORDS="~amd64"` mean that the package can only work on 64-bits? As I'm not that familiar with ebuilds, is there a reason for that? Can it be updated to include keywords for other architectures that are normally supported (so, at least x86, arm and arm64)?

As far as I know, its just to indicate which platforms are supported. So yes it can be updated to include `KEYWORDS="~amd64 ~arm ~arm64 ~x84"` if you support them.

**Reply by filips123** [reply]
I now added automatic publishing with PRs. I created [a template ebuild](https://github.com/filips123/PWAsForFirefox/blob/main/native/packages/gentoo/firefoxpwa.ebuild) that is then updated with `pycargoebuild` to include all crates and then pushed to a PR. So, if you want to change anything in the ebuild in the future, please create a PR with those changes to this repo.

However, publishing didn't work for 2.8.0 because I configured something wrong, so I will manually publish it. I think I fixed this, so it should work fine for future versions.
